### DISCUSSION TITLE
Better documentation for NativeMethods.txt

### DISCUSSION THREAD

**Discussion by DerekZiemba** [microsoft/CsWin32#428] [Q&A]
I can't find an example of a NativeMethods.txt file that actually implements what the documentation says is possible. 

>Create a `NativeMethods.txt` file in your project directory that lists the APIs to generate code for.
Each line may consist of *one* of the following:
> * Exported method name (e.g. `CreateFile`). This *may* include the `A` or `W` suffix, where applicable.
> * A namespace to generate all APIs within (e.g. `Windows.Win32.Storage.FileSystem`).
> * Module name followed by `.*` to generate all methods exported from that module (e.g. `Kernel32.*`).
> * The name of a struct, enum, constant or interface to generate.
> * A comment (i.e. any line starting with `//`) or white space line, which will be ignored.

What's the full syntax to actually achieve all that?

For example, the following does not work but the documentation above leads me to believe it should:
```
Kernel32.GetConsoleWindow
Kernel32.CreateMutex
Kernel32.ReleaseMutex
User32.ShowWindow
User32.MessageBox
```

It also says I can specify a namespace to generate the APIs within, how exactly would that be achieved?

I realize it works if I remove the the module name. But what if there's two methods with the same name in two modules?  How would that work?

**Comment by jnm2** [comment]
The documentation says "Each line may consist of _one_ of the following," so that would say that `GetConsoleWindow` works and `Kernel32.*` works but leads me to believe that `Kernel32.GetConsoleWindow` would not work, for instance.

> I realize it works if I remove the the module name. But what if there's two methods with the same name in two modules? How would that work?

Does a collision between method names ever happen in Win32 API? If so, how would one deal with this when using such APIs from C++?

**Reply by DerekZiemba** [reply]
What about the namespace to generate all APIs in?

Or is it saying it will generate all API's *from* a namespace?

**Reply by AArnott** [reply]
If a collision between two methods exists (very rare), then CsWin32 will direct you to qualify the method with a namespace (not the module name). It will provide a list of all the fully-qualified method names that could match so that you can easily update your NativeMethods.txt file with copy-paste.

> What about the namespace to generate all APIs in? Or is it saying it will generate all API's from a namespace?

I see how that is ambiguous. I will fix.
